full_storage_utilization_test.py

```diff
@@ -0,0 +1,146 @@
+
+    def start_throttle_read(self):
+        self.stress_cmd_r = self.stress_cmd_r.replace("<THREADS_PLACE_HOLDER>", str(self.num_stress_threads))
+        self.run_stress_thread(stress_cmd=self.stress_cmd_r)
+        time.sleep(self.sleep_time_before_scale)
+
+    def start_throttle_rw(self):
+        self.start_throttle_write()
+        self.start_throttle_read()
+        
+    def scale_out(self):
+        self.start_throttle_rw()
+        self.log.info("Started adding a new node")
+        start_time = time.time()
+        self.add_new_node()
+        duration = time.time() - start_time
+        self.log.info(f"Adding a node finished with time: {duration}")
+
+    def scale_in(self):        
+        self.start_throttle_rw()
+        self.log.info("Started removing a node")
+        start_time = time.time()
+        self.remove_node()
+        duration = time.time() - start_time
+        self.log.info(f"Removing a node finished with time: {duration}")
+
+    def drop_data(self, keyspace_name):
+        '''
+        Drop keyspace and clear snapshots.
+        '''
+        node = self.db_cluster.nodes[0]
+        self.log.info("Dropping some data")
+        query = f"DROP KEYSPACE {keyspace_name}"
+        with self.db_cluster.cql_connection_patient(node) as session:
+           session.execute(query)
+           #node.run_nodetool(f"clearsnapshot")
+        DiskUtils.log_disk_usage(self.db_cluster.nodes)     
+
+    def perform_scale_in(self):
+        '''
+        If we are already at 90% disk utilization, first scale-out then
+        drop 20% of data to make space for scale-in operation.
+
+        If we are at 67% disk utilization, scale-in without scale-out or
+        dropping data.
+        '''
+        if self.hardlimit == 90:
+           self.scale_out()           
+           '''
+           Before removing a node, we should make sure
+           other nodes has enough space so that they
+           can accommodate data from the removed node.
+           '''
+           # Remove 20% of data from the cluster.
+           self.drop_data("keyspace_large1")
+           self.drop_data("keyspace_large2")
+           self.scale_in()
+        elif self.hardlimit == 67:
+           self.scale_in()
+
+    def perform_action(self):
+        DiskUtils.log_disk_usage(self.db_cluster.nodes)
+        # Trigger specific action
+        if self.scaling_action_type == ScalingActionType.SCALE_OUT.value:
+           self.scale_out()
+        elif self.scaling_action_type == ScalingActionType.SCALE_IN.value:
+           self.perform_scale_in()
+        else:
+           self.log.info(f"Invalid ActionType {self.scaling_action_type}")
+        DiskUtils.log_disk_usage(self.db_cluster.nodes)
+
+    def test_storage_utilization(self):
+        """
+        Write data until 90% disk usage is reached.
+        Sleep for 60 minutes.
+        Perform specific action.
+        """
+        self.run_stress(self.softlimit, sleep_time=self.sleep_time_fill_disk)
```

Comment:
From the test runs,  having `sleep` after softlimit before hardlimit helps to reach 90% more smoothly. Otherwise test ran into issues due to compaction.